pkg/cmd/pulumi/preview.go

```diff
@@ -18,22 +18,152 @@ import (
 	"bytes"
 	"errors"
 	"fmt"
+	"os"
 
 	"github.com/spf13/cobra"
 
 	"github.com/pulumi/pulumi/pkg/v3/backend"
 	"github.com/pulumi/pulumi/pkg/v3/backend/display"
 	"github.com/pulumi/pulumi/pkg/v3/engine"
 	"github.com/pulumi/pulumi/pkg/v3/resource/deploy"
+	"github.com/pulumi/pulumi/pkg/v3/resource/deploy/providers"
 	"github.com/pulumi/pulumi/pkg/v3/resource/stack"
 	"github.com/pulumi/pulumi/sdk/v3/go/common/apitype"
 	"github.com/pulumi/pulumi/sdk/v3/go/common/diag"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/promise"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/resource"
 	"github.com/pulumi/pulumi/sdk/v3/go/common/util/cmdutil"
 	"github.com/pulumi/pulumi/sdk/v3/go/common/util/contract"
 	"github.com/pulumi/pulumi/sdk/v3/go/common/util/result"
 	"github.com/pulumi/pulumi/sdk/v3/go/common/workspace"
 )
 
+// buildImportFile takes an event stream from the engine and builds an import file from it for every create.
+func buildImportFile(events <-chan engine.Event) *promise.Promise[importFile] {
+	return promise.Run(func() (importFile, error) {
+		// A mapping of every URN we see to it's name, used to build later resourceSpecs
+		fullNameTable := map[resource.URN]string{}
+		// A set of all URNs we've added to the import list, used to avoid adding parents to NameTable.
+		importSet := map[resource.URN]struct{}{}
+		// All providers that we've seen so far, used to build Version and PluginDownloadURL.
+		providerInputs := map[resource.URN]resource.PropertyMap{}
+
+		imports := importFile{
+			NameTable: map[string]resource.URN{},
+		}
+
+		// This is a pretty trivial mapping of Create operations to import declarations.
+		for e := range events {
+			preEvent, ok := e.Payload().(engine.ResourcePreEventPayload)
+			if !ok {
+				continue
+			}
+
+			urn := preEvent.Metadata.URN
+			fullNameTable[urn] = urn.Name()
+
+			// If this is a provider we need to note we've seen it so we can build the Version and PluginDownloadURL of
+			// any resources that use it.
+			if providers.IsProviderType(urn.Type()) {
+				providerInputs[urn] = preEvent.Metadata.Res.Inputs
+			}
+
+			// Only interested in creates
+			if preEvent.Metadata.Op != deploy.OpCreate {
+				continue
+			}
```

Comment:
So this only works on fresh (empty) stacks? I can't generate an import file based on an existing stack with already created resources?